src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +108,130 @@ def self.compile_ast(code, compile_options, boltlib_path)
               compiler.compile_additions
               compiler.catalog_data_hash
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # When we do not need to load bolt we assume there are no Bolt types to resolve in
+        # AST compilation.
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+      def self.with_puppet_settings(overrides = {})
+        # Create a copy of the existing settings object
+        original_settings = Puppet.settings
+
+        Dir.mktmpdir('bolt_compile') do |dir|
+          puppet_config = {}
+          Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+            puppet_config[setting] = dir
+          end
+          # Replace the existing puppet settings object with one that has 'empty' configuration
+          # (i.e. configuration pointed at an empty dir where puppet won't load things like
+          # puppet.conf)
+          Puppet.replace_settings_object(Puppet::Server::Settings.new(global_settings: [],
+                                                                      puppet_config: puppet_config))
+
+          Puppet.settings.generate_new_local_settings
+
+          # Override module locations, Bolt includes vendored modules in its internal modulepath.
+          Puppet.settings.override_default(:basemodulepath, '')
+          Puppet.settings.override_default(:vendormoduledir, '')
+
+          overrides.each do |setting, value|
+            Puppet.settings[setting] = value
+          end
+
+          yield
+
+          # Return settings to their original form
+          Puppet.replace_settings_object(original_settings)
```

Comment:
So it appears that now this results in a settings object being returned (instead of the compiled catalog) https://github.com/puppetlabs/puppetserver/pull/2555/commits/3f8239405c58a397435cb254a95f513bd45beae2#diff-6f7f4ebbd198c182ba9b89fe43839b117bc1899b5bb3f7ffeddcd1cc64cd79abR95 

I think instead of doing work after the `yield` we just want to use `ensure`. This will allow the catalog to still be bubbled up from this block and ensure that we always redo the settings.

Reply:
Oh, yeah heh